ENH: restore unpack bit lookup table

juliantaylor opened the pull request:
Restores the good performance of unpackbits we had since 1.13.

Closes gh-13537

juliantaylor committed:
ENH: restore unpack bit lookup table

Restores the good performance of unpackbits we had since 1.13.
Added a second lookup table for the new little bitorder flag and changed
it to be stored in little endian as it is the more common endian.

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1788,6 +1791,35 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
+            /* for bitorder little the lookup table is just byte swapped */
+            v_l = npy_bswap8(v_b);
+
+            /* the table is stored as big endian, swap the native integer */
```
it is probably a little odd that we store the table in big endian as the most common platform is little endian
it was like that originaly as the packbit worked in "big endian" bits, now that it can do both should we change the lookup table to be stored in little endian?

eric-wieser replied:
+1 from me

eric-wieser replied:
Wait, isn't the lookup table already little endian, with the low-order bits at the lowest byte of `v_b`?

juliantaylor replied:
little and big endian are a bit confusing in this context, what is relevant is that the byte pattern of the lookup table is the _same_ on both endianess platforms.
endianess is only relevant when interpreting the bytes as integer which is not done by this function it only extends a byte to 8 bytes and integer is simply the datatype used to construct the bytes.
I'll change the comments to say fixed endian.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1703,6 +1703,9 @@ pack_bits(PyObject *input, int axis, char order)
 {
+    static int unpack_init = 0;
+    static npy_uint64 unpack_lookup_l[256];
+    static npy_uint64 unpack_lookup_b[256];
```
can you remove the endianness dependence by making the definition `npy_uint8[256][8]`? Does that have any effect on performance?

juliantaylor replied:
yes that could be done, but lookup table construction then becomes two looks instead of one. Might be less readable.

juliantaylor replied:
though removing endian mentions from the code might be worth it
I'll update.

juliantaylor replied:
done

juliantaylor committed:
MAINT: convert lookup table generation to byte wise loop

That way we do not need to take care of endianess when creating the
table via integers.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1788,6 +1791,24 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
+            for (k=0; k < 8; k++) {
+                npy_uint8 v = (j & (1 << k)) == (1 << k);
+                unpack_lookup_b[j][7 - k] = v;
+                /* for bitorder little the lookup table is byte swapped */
```
I'd be tempted to attach this comment to bitorder big, since that's the one with the more complex expression

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1703,6 +1703,9 @@ pack_bits(PyObject *input, int axis, char order)
 unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
 {
+    static int unpack_init = 0;
+    static npy_uint8 unpack_lookup_l[256][8];
```
Valuable to tell GCC to align these as 8?

juliantaylor replied:
gcc should do that by itself, at least on 64 bit platforms
also if it does not, we are using memcpy to access this on a 8 byte boundary (so we don't need to care about output alignment) so it won't break even if it is not.

one could enforce it portably be making this a int64 variable as before and casting back to char in the init loop, should we do that?

eric-wieser replied:
I'm worried only about performance here, I agree it'll do the right thing with or without alignment.

I don't think the c standard requires GCC to align this, but perhaps it's clever enough to anyway. You could perhaps declare it an array of unions of u8[8] and u64, which would also allow you to construct one table from the other using bswap, and is a little cleaner that doing the casting you're suggesting.

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1788,6 +1791,24 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
+    if (unpack_init == 0) {
+        npy_uint64 j;
+        for (j=0; j < 256; j++) {
+            npy_uint64 k;
```
This can be a uint8, right? Or perhaps a intp, since it's an array index. Not that it really matters...

eric-wieser commented:
Looks a lot clear to me as the array, thanks!

hameerabbasi commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1788,6 +1791,24 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
+            npy_uint64 k;
+            for (k=0; k < 8; k++) {
+                npy_uint8 v = (j & (1 << k)) == (1 << k);
+                unpack_lookup_b[j][7 - k] = v;
```
Is it possible to detect endian-ness and initialize just one table, or do we need both on a given platform?

hameerabbasi replied:
Also, is it possible to change the access rather than changing the table? Might be a bit more maintainable.

eric-wieser replied:
No, we need both to support the two modes of unpack bits. Platform endianness is irrelevant.

eric-wieser replied:
Creating two tables is done so both can be memcopy'd

juliantaylor replied:
one could byteswap from one table, but that is about 60% slower than a second table.
two extra kb of memory usage should not really matter for typical numpy applications though as the function is not that important and the slowdown would be in the new option only it could be considered

juliantaylor replied:
updated to do that, a bit slower but probably fine

hameerabbasi replied:
The comment about changing access still holds.

hameerabbasi reviewed (approved):
Looks good to me overall, some minor changes!

juliantaylor committed:
MAINT: only use one lookuptable

The little bitorder can be handled by byteswapping the bitorder big
lookuptable.
This is up to 60% slower for the new function but saves some memory for
probably not very much used functions.

juliantaylor committed:
DOC: add new un/packbits signature to documentation

mattip reviewed (approved):
Thanks for un-reverting these changes and even improving on them. IMO this is about as far as we want to go with this now

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1810,39 +1831,75 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
-                for (i = 0; i < 8; i++) {
-                    *outptr = ((*inptr & (128 >> i)) != 0);
-                    outptr += out_stride;
+        if (out_stride == 1) {
```
Aside: If you're feeling adventurous - you might be able to make a similar optimization for `out_stride == -1`

juliantaylor replied:
I'm not sure that can even happen as the output is allocated by the same function and should not have negative strides
but even if its probably unnecessary to have that little extra in the code, use cases are unlikely

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1810,39 +1831,75 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
+            }
+            else {
+                for (index = 0; index < in_n; index++) {
+                    npy_uint64 v = unpack_lookup.uint64[*inptr];
```
With the `union` change above, this becomes `unpack_lookup_big[*inptr].uint64`

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1703,6 +1703,11 @@ pack_bits(PyObject *input, int axis, char order)
+    static int unpack_init = 0;
+    static union {
+        npy_uint8  bytes[256][8];
+        npy_uint64 uint64[256];
```
makes sense, done

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1703,6 +1703,11 @@ pack_bits(PyObject *input, int axis, char order)
+    static union {
+        npy_uint8  bytes[256][8];
+        npy_uint64 uint64[256];
+    } unpack_lookup;
```
done

mhvk commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1810,39 +1835,75 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
-                for (i = 0; i < 8; i++) {
+            if (order == 'b') {
+                for (index = 0; index < in_n; index++) {
+                    for (i = 0; i < 8; i++) {
```
Wouldn't it make sense to still use the lookup table even for stride != 1 (i.e., replace the `memcpy`, not the table reading).

juliantaylor replied:
probably, I never bothered testing as it is probably an edge case, though as we are updating it we might as well go all the way

mattip replied:
If the proposed changes will take more than a day or two, perhaps we should merge this as-is with a comment for future improvements, since the deadline for 1.17 is soon

juliantaylor replied:
I probably won't get to do it, imo this can be merged now

seberg replied:
Anything holding this up? Otherwise I will make a pass and merge soon (or probably Matti beats me to it ;)).

charris replied:
LGTM. The suggestion to use the table for the strided case is good, but can go in later, we need the fix for 1.17.
